Tests/SwiftFusionTests/Inference/PPCATrackingFactorTests.swift

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import TensorFlow
+import XCTest
+
+import PenguinStructures
+@testable import SwiftFusion
+
+class PPCATrackingFactorTests: XCTestCase {
+  /// Tests that the custom `linearized(at:)` method produces the same results at autodiff.
+  func testLinearizedValue() {
+    // Initialize `factor` with random fields.
+    let image = Tensor<Double>(randomNormal: [500, 500, 1])
+    let W = Tensor<Double>(
+      randomNormal: PPCATrackingFactor.Patch.shape + [PPCATrackingFactor.V1.dimension])
```

Comment:
I know this PR is about the speedup, but since I'm reading to code in detail (and it has been a while) allow me to spam it with general thoughts about the overall design. 

In particular: It seems odd to have to relate to a *factor* here. We might want to refactor (no pun intended!) so we have a *PPCA generative model* that is separate from the factor, which would use it. That sets up a good pattern for the VAE iteration.

Reply:
Could you say more about what seems odd to you here, Frank?

Reply:
I see that templating it could be good.

I think we should start out doing copy and paste though because there are some pieces that are not completely trivial to abstract over (like custom jacobian) and it'll be easier to abstract over them once we have a few examples.